format/spec.md

```diff
@@ -230,11 +233,31 @@ Schemas may be evolved by type promotion or adding, deleting, renaming, or reord
 
 Evolution applies changes to the table's current schema to produce a new schema that is identified by a unique schema ID, is added to the table's list of schemas, and is set as the table's current schema.
 
-Valid type promotions are:
+Valid primitive type promotions are:
+
+| Primitive type   | v1, v2 valid type promotions | v3+ valid type promotions    | Requirements |
+|------------------|------------------------------|------------------------------|--------------|
+| `unknown`        |                              | _any type_                   | |
+| `int`            | `long`                       | `long`                       | |
+| `date`           |                              | `timestamp`, `timestamp_ns`  | Promotion to `timestamptz` or `timestamptz_ns` is **not** allowed |
```

Comment:
Should we specify how overflow should be handled for date->timestamp_ns ?

Reply:
I think to timestamp isn't an issue as long as we assume dates are in a reasonable range (e.g. years 0001 and 9999)

Reply:
Yes, we should. Should it fail or result in null?

Reply:
After thinking about this, I'm going to update it to require failure when there is an overflow. That doesn't introduce any issues with required fields, where engines may not be expecting a null value.

Reply:
The other option would be return the MIN/MAX value as appropriated depending on the which side the date is out of range.  I agree returning an error is probably the least likely to cause bugs (I haven't through through all implications of mutating values).

I agree Null's is not viable, in addition to required columns, it potentially breaks ordering since it would conflate overflow in both directions into the same class.

Reply:
Updated.